# Student grades prediction

here all the library.

In [204]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.svm import SVR
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import cross_val_score, GridSearchCV


import warnings
warnings.filterwarnings("ignore")

In [26]:
data= pd.read_csv("Grades.csv")
data

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0    CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1    CS-97002      A      D     D+         D     B-      C      D      A   
2    CS-97003      A      B      A        B-     B+      A     B-     B+   
3    CS-97004      D     C+     D+         D      D     A-     D+     C-   
4    CS-97005     A-     A-     A-        B+      A      A     A-     B+   
..        ...    ...    ...    ...       ...    ...    ...    ...    ...   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
0       B-  ...     C-     C-     C-     C-     A-      A     C-      B   
1       D+  ...     D+      D      C      D     A-     B-      C      C   
2       A-  ...      B      B      A      C      A      A      A     A-   
3        D  ...     D+      C     D+     C-     B-      B     C+     C+   
4        A  ...     B-     B+     B+     B-     A-      A     A-     A-   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
566     A+  ...     A-     A-      A      A      A     B+     B+      B   
567      A  ...     B+     B+      A      A     A-      B     A-      C   
568      A  ...     A-      B      A     B+      A      C     B+     A-   
569     B-  ...      D      B      B     C-      D      C      B     B-   
570      C  ...     C+      C     B-      D      F     C-     B+      D   

    CS-412   CGPA  
0       A-  2.205  
1        B  2.008  
2        A  3.608  
3       C+  1.906  
4        A  3.448  
..     ...    ...  
566      A  3.798  
567     A-  3.772  
568     A-  3.470  
569      C  2.193  
570     C-  1.753  

[571 rows x 43 columns]

# cheking first five rows of the dataset-

In [27]:
data.head()

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102 EE-119  \
0  CS-97001     B-     D+     C-         C     C-     D+      D     C-     B-   
1  CS-97002      A      D     D+         D     B-      C      D      A     D+   
2  CS-97003      A      B      A        B-     B+      A     B-     B+     A-   
3  CS-97004      D     C+     D+         D      D     A-     D+     C-      D   
4  CS-97005     A-     A-     A-        B+      A      A     A-     B+      A   

   ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423 CS-412   CGPA  
0  ...     C-     C-     C-     C-     A-      A     C-      B     A-  2.205  
1  ...     D+      D      C      D     A-     B-      C      C      B  2.008  
2  ...      B      B      A      C      A      A      A     A-      A  3.608  
3  ...     D+      C     D+     C-     B-      B     C+     C+     C+  1.906  
4  ...     B-     B+     B+     B-     A-      A     A-     A-      A  3.448  

[5 rows x 43 columns]

# cheking first five rows of the dataset-

In [28]:
data.tail()

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
566     A+  ...     A-     A-      A      A      A     B+     B+      B   
567      A  ...     B+     B+      A      A     A-      B     A-      C   
568      A  ...     A-      B      A     B+      A      C     B+     A-   
569     B-  ...      D      B      B     C-      D      C      B     B-   
570      C  ...     C+      C     B-      D      F     C-     B+      D   

    CS-412   CGPA  
566      A  3.798  
567     A-  3.772  
568     A-  3.470  
569      C  2.193  
570     C-  1.753  

[5 rows x 43 columns]

In [137]:
# Drop irrelevant columns like Seat No
df = df.drop(columns=["Seat No."])

In [155]:
# Identify columns containing string values (assuming all columns except CGPA are grades)
grade_cols = df.columns.drop("CGPA")


In [156]:
# Encode categorical variables
label_encoder = LabelEncoder()
for col in grade_cols:
    df[col] = label_encoder.fit_transform(df[col])

In [157]:
# Split the data into features (grades) and target variable (CGPA)
X = df.drop(columns=["CGPA"])
y = df["CGPA"]

# Exploratory Data Analysis (EDA)

# Checking columns

In [158]:
data.columns

Index(['Seat No.', 'PH-121', 'HS-101', 'CY-105', 'HS-105/12', 'MT-111',
       'CS-105', 'CS-106', 'EL-102', 'EE-119', 'ME-107', 'CS-107', 'HS-205/20',
       'MT-222', 'EE-222', 'MT-224', 'CS-210', 'CS-211', 'CS-203', 'CS-214',
       'EE-217', 'CS-212', 'CS-215', 'MT-331', 'EF-303', 'HS-304', 'CS-301',
       'CS-302', 'TC-383', 'MT-442', 'EL-332', 'CS-318', 'CS-306', 'CS-312',
       'CS-317', 'CS-403', 'CS-421', 'CS-406', 'CS-414', 'CS-419', 'CS-423',
       'CS-412', 'CGPA'],
      dtype='object')

In [159]:
data.shape

(571, 43)

In [160]:
df= data.drop_duplicates()
print("Rows & Columns after dropping duplicate : ", df.shape)

Rows & Columns after dropping duplicate :  (571, 43)


Not any duplicates values present in this datasets.

In [161]:
df["PH-121"].value_counts()

A-    112
A     111
B+     61
B      57
B-     56
D      44
C      33
C+     31
D+     22
A+     22
C-     19
WU      2
F       1
Name: PH-121, dtype: int64

In [162]:
df["HS-101"].value_counts()

A-    82
B-    78
C     68
B     63
B+    59
C-    50
C+    47
D     45
A     38
D+    36
A+     4
F      1
Name: HS-101, dtype: int64

# information of dataset-

In [163]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 571 entries, 0 to 570
Data columns (total 43 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Seat No.   571 non-null    object 
 1   PH-121     571 non-null    object 
 2   HS-101     571 non-null    object 
 3   CY-105     570 non-null    object 
 4   HS-105/12  570 non-null    object 
 5   MT-111     569 non-null    object 
 6   CS-105     571 non-null    object 
 7   CS-106     569 non-null    object 
 8   EL-102     569 non-null    object 
 9   EE-119     569 non-null    object 
 10  ME-107     569 non-null    object 
 11  CS-107     569 non-null    object 
 12  HS-205/20  566 non-null    object 
 13  MT-222     566 non-null    object 
 14  EE-222     564 non-null    object 
 15  MT-224     564 non-null    object 
 16  CS-210     564 non-null    object 
 17  CS-211     566 non-null    object 
 18  CS-203     566 non-null    object 
 19  CS-214     565 non-null    object 
 20  EE-217    

using this information we can 1 column is float data types and 42 columns are object data types.

In [164]:
df.isnull().sum()

Seat No.      0
PH-121        0
HS-101        0
CY-105        1
HS-105/12     1
MT-111        2
CS-105        0
CS-106        2
EL-102        2
EE-119        2
ME-107        2
CS-107        2
HS-205/20     5
MT-222        5
EE-222        7
MT-224        7
CS-210        7
CS-211        5
CS-203        5
CS-214        6
EE-217        6
CS-212        6
CS-215        6
MT-331        9
EF-303       10
HS-304       10
CS-301       10
CS-302       10
TC-383       10
MT-442       10
EL-332        9
CS-318        9
CS-306        9
CS-312       10
CS-317       12
CS-403       12
CS-421       12
CS-406       85
CS-414       13
CS-419       13
CS-423       14
CS-412       79
CGPA          0
dtype: int64

we have some null values present in the dataset.

In [165]:
df.describe()

CGPA
count  571.000000
mean     2.954888
std      0.620552
min      0.800000
25%      2.538000
50%      3.029000
75%      3.451000
max      3.985000

Using the describe method to cheque the float data details there are only one columns which have floating value in them and it looks like the count means standard derivation, minimum values 25% quartile, 50% quartile, 75% quartile and maximum value are all the property distributed in term of data points showing on the abnormality as such.

In [166]:
df.describe(include="all").T

count unique       top freq      mean       std  min    25%    50%  \
Seat No.     571    571  CS-97001    1       NaN       NaN  NaN    NaN    NaN   
PH-121       571     13        A-  112       NaN       NaN  NaN    NaN    NaN   
HS-101       571     12        A-   82       NaN       NaN  NaN    NaN    NaN   
CY-105       570     13         A  177       NaN       NaN  NaN    NaN    NaN   
HS-105/12    570     13         A   96       NaN       NaN  NaN    NaN    NaN   
MT-111       569     13        A-  105       NaN       NaN  NaN    NaN    NaN   
CS-105       571     11         A  151       NaN       NaN  NaN    NaN    NaN   
CS-106       569     13        A-  116       NaN       NaN  NaN    NaN    NaN   
EL-102       569     13        A-  105       NaN       NaN  NaN    NaN    NaN   
EE-119       569     12        A-  137       NaN       NaN  NaN    NaN    NaN   
ME-107       569     13        A-   79       NaN       NaN  NaN    NaN    NaN   
CS-107       569     14         A  105       NaN       NaN  NaN    NaN    NaN   
HS-205/20    566     13        A-  150       NaN       NaN  NaN    NaN    NaN   
MT-222       566     14        A-   86       NaN       NaN  NaN    NaN    NaN   
EE-222       564     13         A  122       NaN       NaN  NaN    NaN    NaN   
MT-224       564     14        A-  120       NaN       NaN  NaN    NaN    NaN   
CS-210       564     14        A-  133       NaN       NaN  NaN    NaN    NaN   
CS-211       566     14        A-   68       NaN       NaN  NaN    NaN    NaN   
CS-203       566     13        A-   88       NaN       NaN  NaN    NaN    NaN   
CS-214       565     13         C   76       NaN       NaN  NaN    NaN    NaN   
EE-217       565     13        A-  137       NaN       NaN  NaN    NaN    NaN   
CS-212       565     12        A-  101       NaN       NaN  NaN    NaN    NaN   
CS-215       565     14        A-   79       NaN       NaN  NaN    NaN    NaN   
MT-331       562     12         A  118       NaN       NaN  NaN    NaN    NaN   
EF-303       561     12         B  112       NaN       NaN  NaN    NaN    NaN   
HS-304       561     14        A-  128       NaN       NaN  NaN    NaN    NaN   
CS-301       561     12        A-  108       NaN       NaN  NaN    NaN    NaN   
CS-302       561     11        A-  113       NaN       NaN  NaN    NaN    NaN   
TC-383       561     12         A  105       NaN       NaN  NaN    NaN    NaN   
MT-442       561     12        A-  140       NaN       NaN  NaN    NaN    NaN   
EL-332       562     13        A-   96       NaN       NaN  NaN    NaN    NaN   
CS-318       562     14        A-   89       NaN       NaN  NaN    NaN    NaN   
CS-306       562     13        A-  120       NaN       NaN  NaN    NaN    NaN   
CS-312       561     14        A+   93       NaN       NaN  NaN    NaN    NaN   
CS-317       559     12        B-   79       NaN       NaN  NaN    NaN    NaN   
CS-403       559     11         A  133       NaN       NaN  NaN    NaN    NaN   
CS-421       559     13         B   86       NaN       NaN  NaN    NaN    NaN   
CS-406       486     14        A-  177       NaN       NaN  NaN    NaN    NaN   
CS-414       558     13         A  176       NaN       NaN  NaN    NaN    NaN   
CS-419       558     12        A-  120       NaN       NaN  NaN    NaN    NaN   
CS-423       557     12        A-  122       NaN       NaN  NaN    NaN    NaN   
CS-412       492     13        A-  157       NaN       NaN  NaN    NaN    NaN   
CGPA       571.0    NaN       NaN  NaN  2.954888  0.620552  0.8  2.538  3.029   

             75%    max  
Seat No.     NaN    NaN  
PH-121       NaN    NaN  
HS-101       NaN    NaN  
CY-105       NaN    NaN  
HS-105/12    NaN    NaN  
MT-111       NaN    NaN  
CS-105       NaN    NaN  
CS-106       NaN    NaN  
EL-102       NaN    NaN  
EE-119       NaN    NaN  
ME-107       NaN    NaN  
CS-107       NaN    NaN  
HS-205/20    NaN    NaN  
MT-222       NaN    NaN  
EE-222       NaN    NaN  
MT-224       NaN    NaN  
CS-2

In [167]:
df.nunique().to_frame("Unique Values")

Unique Values
Seat No.             571
PH-121                13
HS-101                12
CY-105                13
HS-105/12             13
MT-111                13
CS-105                11
CS-106                13
EL-102                13
EE-119                12
ME-107                13
CS-107                14
HS-205/20             13
MT-222                14
EE-222                13
MT-224                14
CS-210                14
CS-211                14
CS-203                13
CS-214                13
EE-217                13
CS-212                12
CS-215                14
MT-331                12
EF-303                12
HS-304                14
CS-301                12
CS-302                11
TC-383                12
MT-442                12
EL-332                13
CS-318                14
CS-306                13
CS-312                14
CS-317                12
CS-403                11
CS-421                13
CS-406                14
CS-414                13
CS-419                12
CS-423                12
CS-412                13
CGPA                 491

With the uses of any nunique method we can see the total number of unique values present in all the columns.

# splitting the dataset into two variables namely X and Y for feature and lable:

In [168]:
x=df.drop("CGPA", axis=1)
y=df["CGPA"]

In [169]:
x

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0    CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1    CS-97002      A      D     D+         D     B-      C      D      A   
2    CS-97003      A      B      A        B-     B+      A     B-     B+   
3    CS-97004      D     C+     D+         D      D     A-     D+     C-   
4    CS-97005     A-     A-     A-        B+      A      A     A-     B+   
..        ...    ...    ...    ...       ...    ...    ...    ...    ...   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-306 CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419  \
0       B-  ...      C     C-     C-     C-     C-     A-      A     C-   
1       D+  ...      D     D+      D      C      D     A-     B-      C   
2       A-  ...     A-      B      B      A      C      A      A      A   
3        D  ...     C-     D+      C     D+     C-     B-      B     C+   
4        A  ...     A-     B-     B+     B+     B-     A-      A     A-   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
566     A+  ...     B+     A-     A-      A      A      A     B+     B+   
567      A  ...     A-     B+     B+      A      A     A-      B     A-   
568      A  ...     B+     A-      B      A     B+      A      C     B+   
569     B-  ...     B+      D      B      B     C-      D      C      B   
570      C  ...     D+     C+      C     B-      D      F     C-     B+   

    CS-423 CS-412  
0        B     A-  
1        C      B  
2       A-      A  
3       C+     C+  
4       A-      A  
..     ...    ...  
566      B      A  
567      C     A-  
568     A-     A-  
569     B-      C  
570      D     C-  

[571 rows x 42 columns]

In [170]:
y

0      2.205
1      2.008
2      3.608
3      1.906
4      3.448
       ...  
566    3.798
567    3.772
568    3.470
569    2.193
570    1.753
Name: CGPA, Length: 571, dtype: float64

# Resolving the class imbalance issue in our label column:

In [171]:
y.value_counts()

3.019    5
3.058    3
2.793    3
3.443    3
2.206    3
        ..
2.555    1
2.042    1
2.634    1
2.053    1
1.753    1
Name: CGPA, Length: 491, dtype: int64

# Scaling the Feature

In [172]:
y_reshaped = y.values.reshape(-1, 1)
y_scaled = scaler.fit_transform(y_reshaped)


In [173]:
y_scaled

array([[-1.20947980e+00],
       [-1.52721731e+00],
       [ 1.05339192e+00],
       [-1.69173115e+00],
       [ 7.95330996e-01],
       [ 1.14695313e-01],
       [ 3.40653976e-03],
       [ 1.42114286e-01],
       [ 6.50171727e-01],
       [ 4.71141962e-01],
       [-4.41748552e-01],
       [ 5.25979908e-01],
       [-2.62718787e-01],
       [ 9.16297054e-01],
       [ 7.87266592e-01],
       [-2.86911998e-01],
       [ 4.72754843e-01],
       [-1.88366396e+00],
       [ 9.95328211e-01],
       [ 8.51781823e-01],
       [-1.14335169e+00],
       [ 1.79365900e-03],
       [-3.04498330e-03],
       [ 1.01790854e+00],
       [ 8.38878777e-01],
       [-1.17238354e+00],
       [ 7.82427950e-01],
       [-1.23689877e+00],
       [ 3.87272162e-01],
       [-1.06593341e+00],
       [ 7.87266592e-01],
       [ 2.87273555e-01],
       [ 4.27594181e-01],
       [ 2.53403059e-01],
       [ 3.30821335e-01],
       [ 1.03405147e-01],
       [ 3.37272858e-01],
       [ 6.08236827e-01],
       [-1.2

In [174]:
df.skew()

CGPA   -0.497203
dtype: float64

# Split the data into training and testing sets

In [209]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# linear regression model

In [210]:
model = LinearRegression()
model.fit(X_train, y_train)

print("Accuracy:", accuracy)

Accuracy: 0.790603838246097


# Decision Tree

In [198]:
model = DecisionTreeRegressor()
model.fit(X_train, y_train)

print("Accuracy:", accuracy)

Accuracy: 0.790603838246097


# SVM

In [211]:
svm_regressor = SVR()
svm_regressor.fit(X_train, y_train)

print("Accuracy:", accuracy)

Accuracy: 0.790603838246097


# Conclusion

Accuracy: 0.790603838246097     all the method predict same Accuracy.